src/pathfinder/pathfinder_func.h

```diff
@@ -31,6 +65,19 @@ inline TileIndex CalcClosestStationTile(StationID station, TileIndex tile, Stati
 	/* If the rail station is (temporarily) not present, use the station sign to drive near the station */
 	if (ta.tile == INVALID_TILE) return st->xy;
 
+	if (station_type == STATION_DOCK) {
+		uint closest_dist = std::numeric_limits<uint>::max();
+		TileIndex closest_tile;
+		for (const TileIndex &docking_tile : ta) {
+			if (!IsDockingTile(docking_tile) || !IsShipDestinationTile(docking_tile, station)) continue;
+			const uint dist = DistanceManhattan(docking_tile, tile);
+			if (dist >= closest_dist) continue;
+			closest_dist = dist;
+			closest_tile = docking_tile;
+		}
+		return closest_tile;
```

Comment:
I've not read the whole PR to know how this works, but by picking the "closest" docking location by direct distance means you can end up picking a location that cannot be reached, or the path to it could be much further than the path to another docking location.